app/src/main/java/com/github/se/icebreakrr/model/notification/EngagementNotificationManager.kt

```diff
@@ -0,0 +1,140 @@
+
+  /** Start monitoring for nearby users with common tags */
+  fun startMonitoring() {
+    stopMonitoring() // Stop any existing monitoring
+
+    notificationJob =
+        scope.launch {
+          while (true) {
+            checkNearbyUsersForCommonTags()
+            delay(CHECK_INTERVAL)
+          }
+        }
+  }
+
+  /** Stop monitoring for nearby users */
+  fun stopMonitoring() {
+    notificationJob?.cancel()
+    notificationJob = null
+  }
+
+  private fun isAppInForeground(): Boolean {
+    val activityManager = context.getSystemService(Context.ACTIVITY_SERVICE) as ActivityManager
+    val appProcesses = activityManager.runningAppProcesses ?: return false
+    val packageName = context.packageName
+
+    for (appProcess in appProcesses) {
+      if (appProcess.importance == ActivityManager.RunningAppProcessInfo.IMPORTANCE_FOREGROUND &&
+          appProcess.processName == packageName) {
+        return true
+      }
+    }
+    return false
+  }
+
+  private fun checkNearbyUsersForCommonTags() {
+    val selfProfile = profilesViewModel.selfProfile.value ?: return
+    val selfLocation = selfProfile.location ?: return
+
+    // Use the user's selected radius from FilterViewModel
+    profilesViewModel.getFilteredProfilesInRadius(
+        center = selfLocation,
+        radiusInMeters = filterViewModel.selectedRadius.value,
+        genders = filterViewModel.selectedGenders.value,
+        ageRange = filterViewModel.ageRange.value)
+
+    // Launch a coroutine to collect the filtered profiles
+    scope.launch {
+      // Only proceed if we are discoverable
+      if (!appDataStore.isDiscoverable.first()) return@launch
+
+      profilesViewModel.filteredProfiles.collectLatest { nearbyProfiles ->
+        // Only send notifications if the app is not in foreground
+        if (!isAppInForeground()) {
+          processNearbyProfiles(selfProfile, nearbyProfiles)
+        }
+      }
+    }
+  }
+
+  private fun processNearbyProfiles(selfProfile: Profile, nearbyProfiles: List<Profile>) {
+    val selfTags = selfProfile.tags
+
+    for (nearbyProfile in nearbyProfiles) {
+      if (nearbyProfile.uid == selfProfile.uid) continue // Skip self
+
+      // Find common tags
+      val commonTags = selfTags.intersect(nearbyProfile.tags.toSet())
+
+      if (commonTags.isNotEmpty()) {
+        // Send notification for the first common tag
+        val commonTag = commonTags.first()
+        try {
+          // Send notification to the other user
+          meetingRequestViewModel.engagementNotification(
+              targetToken = nearbyProfile.fcmToken ?: continue, tag = commonTag)
+
+          // Send notification to self as well
+          meetingRequestViewModel.engagementNotification(
+              targetToken = selfProfile.fcmToken ?: continue, tag = commonTag)
+        } catch (e: Exception) {
+          Log.e("EngagementNotification", "Failed to send notification", e)
+        }
```

Comment:
This part is a good use of my code of the MeetingRequestViewModel, well done ! Just one question tho :
1) Maybe it isn't very clear in my messaging implementation but normally a Profile ll always have an FCM token, because when Signing in, the FCM token is directly added. So i think the "continue" can be replaced by strings like ""null"", like i did in my implementation. What do you think ? The advantage of "null" is that it would trigger errors so it could help for debugging and bad behavior. But the use of "continue " is also correct, but it avoids the errors

Reply:
you're right i'll change it